File: algorithms/Javascript/maxNonAdjacentSubsetSum.js
function maxNonAdjacentSubsetSum(array) {
	if (!array.length) return 0;
	if (array.length === 1) return array[0];
	const maxSums = array.slice();
	maxSums[1] = Math.max(array[0], array[1]);
	for (let i = 2; i < array.length; i++) {
		maxSums[i] = Math.max(maxSums[i - 1], maxSums[i - 2] + array[i]);
	} // take max from last position vs two positions ago plus current num
	return maxSums[maxSums.length - 1];
}

// O(n) Time | O(1) Space
function maxSubsetSumNonAdjacent(array) {
	if (!array.length) return 0;
	if (array.length === 1) return array[0];
	let first = Math.max(array[0], array[1]);
	let second = array[0];
	for (let i = 2; i < array.length; i++) {
		let current = Math.max(first, second + array[i]);
		second = first;
		first = current;
	}
	return first;
}

File: algorithms/Python/maxNonAdjacentSubsetSum.py
# O(n) Time | O(n) Space
def maxNonAdjacentSubsetSum(array):
    if not len(array): # if array is empty
        return 0
    elif len(array) == 1:
        return array[0]
    maxSums = array[:] # copy of whole array, maxSums[0] = array[0]
    maxSums[1] = max(array[0], array[1])
    for i in range(2, len(array)):
        maxSums[i] = max(maxSums[i - 1], maxSums[i - 2] + array[i])
    return maxSums[-1]

# O(n) Time | O(1) Space
def maxSubsetSumNonAdjacent(array):
    if not len(array):
        return 0
    elif len(array) == 1:
        return array[0]
    second = array[0] # represents maxSums[i - 2]
    first = max(array[0], array[1]) # represents maxSums[i - 1]
    for i in range(2, len(array)):
        current = max(first, second + array[i])
        second = first
        first = current
    return first